lib/fuse_lowlevel.c

```diff
@@ -2149,9 +2139,19 @@ void do_init(fuse_req_t req, fuse_ino_t nodeid, const void *inarg)
 		return;
 	}
 
-	if (se->conn.max_write < bufsize - FUSE_BUFFER_HEADER_SIZE) {
-		se->bufsize = se->conn.max_write + FUSE_BUFFER_HEADER_SIZE;
+	if (bufsize < FUSE_MIN_READ_BUFFER) {
+		fuse_log(FUSE_LOG_ERR,
+			 "fuse: warning: buffer size too small: %zu\n",
+			 bufsize);
+		bufsize = FUSE_MIN_READ_BUFFER;
 	}
+
+	if (se->conn.max_write > bufsize - FUSE_BUFFER_HEADER_SIZE)
+		se->conn.max_write = bufsize - FUSE_BUFFER_HEADER_SIZE;
+	if (se->conn.max_write < bufsize - FUSE_BUFFER_HEADER_SIZE)
```

Comment:
Maybe this could be `else if`?

Or, another way of expressing this block of logic that might be simpler:
```
se->bufsize = min(se->conn.max_write, bufsize - FUSE_BUFFER_HEADER_SIZE);
se->conn.max_write = se->bufsize;
```

Reply:
Good idea to further simplify it, I'm just more tended to 

```
	se->conn.max_write = MIN(se->conn.max_write, bufsize - FUSE_BUFFER_HEADER_SIZE);

        /* the code might want to do larger allocations for the buffer than needed, 
         * to avoid fragmentation and let the system use transparent huge pages.
         */
	se->bufsize = MAX(bufsize, se->conn.max_write + FUSE_BUFFER_HEADER_SIZE);
```

What do you think?

Reply:
Btw, I think your above setting is not correct, as max_write is bufsize - FUSE_BUFFER_HEADER_SIZE;

We could also do

```
se->conn.max_write = MIN(se->conn.max_write, bufsize - FUSE_BUFFER_HEADER_SIZE);
se->bufsize = se->conn.max_write + FUSE_BUFFER_HEADER_SIZE;
```

And then let the actual allocation code round it up to something more THP usable, if it wants to.

Reply:
Oh yeah, se->bufsize needs to include the FUSE_BUFFER_HEADER, nice catch.

```
se->conn.max_write = MIN(se->conn.max_write, bufsize - FUSE_BUFFER_HEADER_SIZE); se->bufsize = se->conn.max_write + FUSE_BUFFER_HEADER_SIZE;
```
looks great to me